sys/linux/socket_nvme_of_tcp.txt

```diff
@@ -0,0 +1,623 @@
+# Copyright 2018 syzkaller project authors. All rights reserved.
+# Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+meta arches["amd64"]
+
+include <linux/nvme-tcp.h>
+include <linux/sizes.h>
+include <linux/nvme.h>
+include <linux/socket.h>
+include <uapi/linux/in.h>
+include <linux/net.h>
+
+#########################################
+### Generic NVMe over TCP ###
+
+nvme_of_msg {
+	pdu	nvme_tcp_pdu
+	payload	optional[buffer[in]]
+} [packed]
+
+nvme_tcp_pdu [
+	icreq	nvme_tcp_icreq_pdu
+	icresp	nvme_tcp_icresp_pdu
+	cmd	nvme_tcp_cmd_pdu
+	rsp	nvme_tcp_rsp_pdu
+	r2t	nvme_tcp_r2t_pdu
+	data	nvme_tcp_data_pdu
+]
+
+type nvme_tcp_hdr_t_const[PDU_TYPE, HLEN_SIZE] {
+	type	const[PDU_TYPE, int8]
+	flags	flags[nvme_tcp_pdu_flags, int8]
+	hlen	HLEN_SIZE
+	pdo	int8
+	plen	int32
```

Comment:
Should it be the length of the command or something? If yes, syzkaller can be [flexible](https://github.com/google/syzkaller/blob/master/docs/syscall_descriptions_syntax.md#length) in terms of the length of what exactly you want to specify.

Reply:
The `nvme_tcp_hdr_t_const` is 8 bytes no matter the message size. However `hlen` is changing depending on the message type, which I am not sure we can know in advance without sending it as a const.

Reply:
Tried `len[parent, int8]`. It outputs `0` instead of the parent size size.
I tried it this way:
```
type nvme_tcp_hdr_t_const[PDU_TYPE] {
	type	const[PDU_TYPE, int8]
	flags	flags[nvme_tcp_pdu_flags, int8]
	hlen	len[parent, int8]
	pdo	int8
	plen	int32
} [packed]

nvme_tcp_hdr_data_pdu_union [
    nvme_tcp_hdr_data_pdu_h2c		nvme_tcp_hdr_t_const[nvme_tcp_h2c_data]
    # All the other union members
]
```

Reply:
> I tried it this way:

Please try once more after you remove that union and include the header directly.

> Re plen -
plen and pdo change depending on the object and sometimes depending on a packet that was sent by server previously in the TCP session. Calculating it should be done by another pseudo-syscall and in my opinion should be considered as future work to make this subsystem support better.

I see, thanks for the clarification!

Reply:
> Please try once more after you remove that union and include the header directly.
Tried it after changing everything from union to actual structs and it still does not work.

Reply:
Hmm, yes, it looks like `parent` would still refer only to the header struct, not to the overall struct.

Let's keep it as is then.

Reply:
Should we mark this conversation as resolved?

Reply:
Yes